rich/markdown.py

```diff
@@ -449,83 +478,109 @@ def __rich_console__(
+
+            entering = token.nesting == 1
+            exiting = token.nesting == -1
+            self_closing = token.nesting == 0
+
+            if node_type == "text":
+                context.on_text(token.content, node_type)
+            elif node_type == "hardbreak":
+                context.on_text("\n", node_type)
             elif node_type == "softbreak":
-                if entering:
-                    context.on_text(" ", node_type)
-            elif node_type == "link":
-                if entering:
-                    link_style = console.get_style("markdown.link", default="none")
-                    if self.hyperlinks:
-                        link_style += Style(link=current.destination)
+                context.on_text(" ", node_type)
+            elif node_type == "link_open":
+                href = str(token.attrs.get("href", ""))
+                link_style = console.get_style("markdown.link", default="none")
+                if self.hyperlinks:
+                    link_style += Style(link=href)
                     context.enter_style(link_style)
                 else:
+                    context.stack.push(Link.create(self, token))
+            elif node_type == "link_close":
+                link_style = console.get_style("markdown.link", default="none")
+                if self.hyperlinks:
                     context.leave_style()
-                    if not self.hyperlinks:
-                        context.on_text(" (", node_type)
-                        style = Style(underline=True) + console.get_style(
-                            "markdown.link_url", default="none"
-                        )
-                        context.enter_style(style)
-                        context.on_text(current.destination, node_type)
-                        context.leave_style()
-                        context.on_text(")", node_type)
-            elif node_type in inlines:
-                if current.is_container():
-                    if entering:
-                        context.enter_style(f"markdown.{node_type}")
-                    else:
-                        context.leave_style()
                 else:
-                    context.enter_style(f"markdown.{node_type}")
-                    if current.literal:
-                        context.on_text(current.literal, node_type)
+                    element: Any = context.stack.pop()
+                    # TODO: I want to encapsulate the stuff below into the Link renderable
+                    #  but can't work out how to do it.
+                    context.enter_style(link_style)
+                    context.on_text(element.text.plain, node_type)
+                    context.leave_style()
+                    context.on_text(" (", node_type)
+                    context.enter_style(link_style)
+                    context.on_text(element.href, node_type)
+                    context.leave_style()
+                    context.on_text(")", node_type)
```

Comment:
This was the part I'm very "meh" about. It's basically a hack. As the TODO comment suggests, I wanted to throw the data into the `Link` renderable when the tag opens, and have it deal with this stuff. However, I can't get the information to render in the right place. Right now, we create a `Link`, and put it on the stack. Then we pop it off here and "manually" render the data - it's misusing the stack really.

Reply:
I think I understand what's going on here. You're inserting content that isn't technically in the document, when hyperlinks are disabled?

I doesn't look all *that* dirty to me. But I think you're right that is should probably be the Link object's responsibility.

We can revisit that tomorrow, and try to figure out a cleaner solution. But if this is what we go for, it's not going to keep me up at night ;)